src/Aspire.Hosting.Keycloak/KeycloakResouceBuilderExtensions.cs

```diff
@@ -0,0 +1,96 @@
+        int? port = null,
+        IResourceBuilder<ParameterResource>? admin = null,
+        IResourceBuilder<ParameterResource>? adminPassword = null)
+    {
+        var passwordParameter = adminPassword?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password");
+
+        var resource = new KeycloakResource(name, admin?.Resource, passwordParameter);
+
+        var keycloak = builder
+            .AddResource(resource)
+            .WithImage(KeycloakContainerImageTags.Image)
+            .WithImageRegistry(KeycloakContainerImageTags.Registry)
+            .WithImageTag(KeycloakContainerImageTags.Tag)
+            .WithHttpEndpoint(port: port, targetPort: DefaultContainerPort)
+            .WithEnvironment(context =>
+            {
+                context.EnvironmentVariables[AdminEnvVarName] = resource.AdminReference;
+                context.EnvironmentVariables[AdminPasswordEnvVarName] = resource.AdminPasswordParameter;
+            });
+
+        if (builder.ExecutionContext.IsRunMode)
+        {
+            keycloak.WithArgs("start-dev");
+        }
+        else
+        {
+            keycloak.WithArgs("start");
+        }
+
+        return keycloak;
+    }
+
+    /// <summary>
+    /// Adds a named volume for the data folder to a Keycloak container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only volume.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<KeycloakResource> WithDataVolume(this IResourceBuilder<KeycloakResource> builder, string? name = null, bool isReadOnly = false)
+        => builder.WithVolume(name ?? VolumeNameGenerator.CreateVolumeName(builder, "data"), "/opt/keycloak/data", isReadOnly);
+
+    /// <summary>
+    /// Adds a reference to a Keycloak resource to the specified resource builder.
+    /// </summary>
+    /// <typeparam name="TResource">The type of the resource builder.</typeparam>
+    /// <param name="builder">The resource builder to add the reference to.</param>
+    /// <param name="keycloakBuilder">The Keycloak resource builder to reference.</param>
+    /// <param name="realm">The realm to reference in the Keycloak server.</param>
+    /// <returns>The resource builder with the added reference.</returns>
+    public static IResourceBuilder<TResource> WithReference<TResource>(
+        this IResourceBuilder<TResource> builder,
+        IResourceBuilder<KeycloakResource> keycloakBuilder,
+        string realm
+    )
+        where TResource : IResourceWithEnvironment
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentNullException.ThrowIfNull(keycloakBuilder);
+
+        builder.WithReference(keycloakBuilder)
```

Comment:
Multiple realm registrations are possible. In this case, we override and use the latest one. Is it intended behavior?

Reply:
Yea, multiple realms are not supported with the proposed implementation.